File: python-api/deploy_the_project.py
from theblockchainapi.developer_program_resource import DeveloperProgramResource
import requests
import json


API_KEY_ID = None
API_SECRET_KEY = None


try:
    assert API_KEY_ID is not None
    assert API_SECRET_KEY is not None
except AssertionError:
    raise Exception("Fill out your API keys!")


RESOURCE = DeveloperProgramResource(
    api_key_id=API_KEY_ID,
    api_secret_key=API_SECRET_KEY
)


def test_deployment(project_id: str, version: str, endpoint_path: str, input_dict: dict):
    print("-" * 20)
    print("Testing Deployment...")
    response = requests.post(
        url=f'https://api.blockchainapi.com/third-party-apis/{project_id}/v{version}/{endpoint_path}',
        data=json.dumps(input_dict),
        headers={
            'APIKeyId': API_KEY_ID,
            'APISecretKey': API_SECRET_KEY
        }
    )
    print(json.dumps(response.json(), indent=4))
    print("-" * 20)


def deploy(project_id: str, binary_file_path: str):
    # This function can be modified for any language.
    # Feel free to see how it works internally.
    # It first pulls the the deployment URL, which is a pre-signed S3 URL.
    # You then upload your binary to this URL.
    # Then the function repeatedly calls `get_project_deployment_status` to check whether it was successfully deployed.
    # The time to upload depends on the size of your binary.
    # The time to deploy should be about 10-20 seconds, once uploaded.
    # If it is longer or never shows as deployed, please contact us.
    RESOURCE.deploy_project(project_id, binary_file_path)


def main():
    # Insert the `project_id` generated after creating the project.
    project_id = None
    # This is the default version when creating a project. If you're using a different version, enter it here.
    version = '0.0.1'
    # Modify the path to the generated binary, if necessary.
    # binary_file_path = 'my-api/main'
    # You can also just use the path to our pre-generated binaries, if you'd like to proceed for testing purposes.
    binary_file_path = 'my-api/main'

    deploy(project_id, binary_file_path)

    stats = RESOURCE.get_project_stats(project_id)
    print("STATS\n", json.dumps(stats, indent=4))

    test_deployment(
        project_id=project_id,
        version=version,
        endpoint_path='ping',
        input_dict=dict()
    )

    test_deployment(
        project_id=project_id,
        version=version,
        # The path to the endpoint we created in `create_the_project.py`.
        endpoint_path='generate/public_key',
        input_dict={
            'starting_with': 'C'
        }
    )

    stats = RESOURCE.get_project_stats(project_id)
    print("STATS\n", json.dumps(stats, indent=4))


if __name__ == '__main__':
    main()

File: node-js-api/deploy_the_project.js
const theblockchainapi = require('theblockchainapi');
const axios = require('axios').default;
var FormData = require('form-data');
const { createReadStream } = require("fs");
const request = require('request');

// Install the API wrapper.
// `npm install theblockchainapi request form-data fs axios`
// node create_the_project.js

const DEFAULT_CLIENT = theblockchainapi.ApiClient.instance;

const APIKeyID = DEFAULT_CLIENT.authentications['APIKeyID'];
APIKeyID.apiKey = 'API-KEY-ID';

const APISecretKey = DEFAULT_CLIENT.authentications['APISecretKey'];
APISecretKey.apiKey = 'API-SECRET-KEY';

const PROJECT_API_INSTANCE = new theblockchainapi.ProjectApi();
const ENDPOINT_API_INSTANCE = new theblockchainapi.EndpointApi();

const HEADERS = {
    'APIKeyId': API_KEY_ID,
    'APISecretKey': API_SECRET_KEY
};


const delay = ms => new Promise(res => setTimeout(res, ms));


const testDeployment = async (projectId, version, endpointPath, inputDict) => {
    console.log("--------------------------------------------------------");
    console.log("Testing Deployment...");
    const url = 'https://api.blockchainapi.com/third-party-apis/' + projectId + '/v' + version + '/' + endpointPath;
    
    // const body = JSON.stringify(input_dict);

    const response = await axios.post(
        url,
        inputDict, 
        { 
            headers: HEADERS 
        }
    );

    console.log(response.data, response.status);
    console.log("--------------------------------------------------------");
}


const deploy = async (projectId, binaryFilePath) => {
    // This function can be modified for any language.
    // Feel free to see how it works internally.
    // It first pulls the the deployment URL, which is a pre-signed S3 URL.
    // You then upload your binary to this URL.
    // Then the function repeatedly calls `get_project_deployment_status` to check whether it was successfully deployed.
    // The time to upload depends on the size of your binary.
    // The time to deploy should be about 10-20 seconds, once uploaded.
    // If it is longer or never shows as deployed, please contact us.
    let url = "https://api.blockchainapi.com/v1/project/" + projectId + "/deploy/url";
    
    const response = await axios.post(
        url,
        {
            // 'platform': 'Linux/aarch64'
            'platform': 'Linux'
        }, 
        { 
            headers: HEADERS 
        }
    );

    const result = response.data;
    
    const fields = result['fields']
    url = result['url']

    const form = new FormData();

    Object.entries(fields).forEach(([field, value]) => {
        form.append(field, value);
    });
    form.append("file", createReadStream(binaryFilePath));

    console.log("Uploading...")

    let uploadPromise = new Promise((resolve, _) => {
        setTimeout( function() {
            form.submit(
                url, 
                (
                    (_, res) => {
                        console.log("Done!")
                        resolve({
                            statusCode: res.statusCode,
                            statusMessage: res.statusMessage
                        })
                    }
                )
            );
        }, 25000)
    })

    const res = await uploadPromise;

    if (res.statusCode == 204) {
        console.log("Upload successful. Now monitoring for successful deployment...")
    } else {
        throw Error("Upload failed... Please report this issue to our team so we can help.\n\n" + res)
    }

    let status;
    // Check status
    while (true) {

        const statusResponse = await axios.post(
            "https://api.blockchainapi.com/v1/project/" + projectId + "/deploy/status",
            {}, 
            { 
                headers: HEADERS 
            }
        );
        status = statusResponse.data;

        if ('error_message' in status) {
            throw Error(status['error_message'])
        }

        console.log(status['status'])
        if (status['status_code'] == 1) {
            break
        } else {
            console.log("Checking... ")
            await delay(5000);
        }

    }
    return status
}


const getStats = async (projectId) => {

    const statsResponse = await axios.get(
        "https://api.blockchainapi.com/v1/project/" + projectId + "/stats",
        { 
            headers: HEADERS 
        }
    );

    return statsResponse.data;
}


const main = async () => {
    // Insert the `project_id` generated after creating the project.
    const projectId = '3uB19klqeLtdqTiNyPMH'; // null;
    // This is the default version when creating a project. If you're using a different version, enter it here.
    const version = '0.0.1';
    // Modify the path to the generated binary, if necessary.
    // binary_file_path = 'my-api/main'
    // You can also just use the path to our pre-generated binaries, if you'd like to proceed for testing purposes.
    const binaryFilePath = 'binaries/main-linux';

    await deploy(projectId, binaryFilePath);

    let stats = getStats(projectId);
    console.log("STATS\n",  JSON.stringify(stats, null, 4));

    testDeployment(
        projectId,
        version,
        'ping',
        {}
    );

    testDeployment(
        projectId,
        version,
        // The path to the endpoint we created in `create_the_project.py`.
        'generate/public_key',
        {
            'startingWith': 'C'
        }
    );
    
    stats = getStats(projectId);
    console.log("STATS\n", JSON.stringify(stats, null, 4));
}


main()
